Openstack UPI docs

tomassedovic opened the pull request:
This adds a document describing how to do UPI on OpenStack. There is no automation yet, all the steps described here are using the standard `openstack` command.

We plan to add automation (using Ansible) in follow-up pull requests.

This is enough to do a full UPI deployment.

tomassedovic commented:
/label platform/openstack

tomassedovic commented:
/label platform/openstack

tomassedovic commented:
I believe this is now ready for comments and reviews.

Things left to do:
* Do all the steps here, validate they still work (I've done that on an earlier edit, not on this one)
* Add links where necessary (e.g. to the IPI docs and the networking architecture, Ignition, CoreOS etc.)
* Add a section describing what UPI is and how it differs from IPI

I would really appreciate if someone else actually followed these steps too. Anything missing/incorrect? Is there anything that could be written more clearly? Any additional info that's missing? Any links to add?

mandre commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,806 @@
+=> 203.0.113.19
+```
+
+**NOTE**: throughout this document, we will use `203.0.113.23` as the public IP address for the OpenShift API endpoint and `203.0.113.19` as the public IP for the ingress (`*.apps`) endpoint.
````
Nice, I see that you're using the IP address range for documentation as specified in RFC 5737 :+1:

tomassedovic replied:
I'd been wondering for a while if there's something like example.com, but for TLDs and IP addresses. Turns out there is \o/

pierreprinetti replied:
I also think that adherence to IETF RFCs is of unquestionable value.

However, while our readers will immediately identify `example.com` as a placeholder, on the other hand `192.0.2.0/24` might be copy-pasted in a production setting.

In order to avoid confusion, I propose to be more explicit about the fact that the IP ranges we use in the examples (192.0.2.0/24 and 203.0.113.0/24, aka TEST-NET-1 and TEST-NET-3) are indeed example IPs and must not be used.

From [RFC 5737](https://tools.ietf.org/html/rfc5737):
> Network operators SHOULD add these address blocks to the list of non-routeable address spaces, and if packet filters are deployed, then this address block SHOULD be added to packet filters.

tomassedovic replied:
Yeah that's a fair point. I'd still like to use the ranges that are actually designed for that purpose rather than actual real ranges that are potentially in use.

That said, adding a warning and making it clear you should pick your own safe range is a completely sensible thing to do.

mandre commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,806 @@
+
+```sh
+$ for index in $(seq 0 2); do
+> MASTER_HOSTNAME="openshift-qlvwv-master-$index\n"
````
`openshift-qlvwv` could to be replaced with `$infraID` or something similar so that we can more easily copy and paste.

tomassedovic replied:
Yeah I thought of that and opted for everything being super explicit. But that does indeed hurt copy/paste (same with the `$` prompt) so I'll fix both.

mandre commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,806 @@
+#### Master Security Group Rules
+
+```sh
+$ openstack security group rule create --description "ICMP" --protocol icmp openshift-qlvwv-master
````
Having the prompt $ char makes it hard to use copy & paste. If you want to keep it there for consistency, how about using heredoc:

```sh
$ bash <<SG_END
openstack security group rule create --description "machine config server" --protocol tcp --dst-port 22623 --remote-ip 192.0.2.0/24 openshift-qlvwv-master
openstack security group rule create --description "SSH" --protocol tcp --dst-port 22 openshift-qlvwv-master
[...]
openstack security group rule create --description "VRRP" --protocol vrrp --remote-ip 192.0.2.0/24 openshift-qlvwv-master
SG_END
```

tomassedovic replied:
Do you think it's okay to keep it everywhere else and just drop the prompt here? Or would you prefer I'd leave it out in every shell session?

tomassedovic replied:
I've updated the SG rules and a few other things that would make copy/pasting really hard (the ignition for loops).

Let me know if you'd prefer to remove the prompt everywhere.

mandre commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,806 @@
+
+### Bootstrap Ignition
+
+The generated boostrap ignition file (`bootstrap.ign`) tends to be quite large (around 300MB -- it contains all the manifests, master and worker ignitions etc.). This is generally too big to be passed to the server directly (the OpenStack Nova user data limit is 64KB).
```
Wow, I assume you meant 300kB

tomassedovic replied:
LOL, indeed :-)

mandre commented:
Looks really solid at first glance. I'll try to run a full end to end UPI deployment following the documentation tomorrow and also compare with the UPI docs for other platforms.

EricArrakis commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,806 @@
+files = ignition['storage'].get('files', []);
+files.append({'path': '/etc/hostname', 'mode': 420, 'contents': {'source': 'data:text/plain;charset=utf-8;base64,' + base64.encodestring('$MASTER_HOSTNAME').strip(), 'verification': {}}, 'filesystem': 'root'});
+ignition['storage']['files'] = files;
+json.dump(ignition, sys.stdout)" <master.ign >"master-$index-ignition.json"
```
Why don't we insert the random generated string (using @mandre 's "$infraID" suggestion) in the name of the master-$index-ignition.json file?  If the admin is setting up multiple clusters, wouldn't this be a more user friendly approach:  they would know which host the IGN file applies to and they don't risk clobbering a previous file?

tomassedovic replied:
These ignition files are stored in a directory that does *not* support multiple deployments. You have to have a separate dir for each deployment.

That said, I've added this anyway to make things more consistent. It's nice to know which ignition to pass to which server since their names match.

EricArrakis commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,806 @@
+If it does (this one is for `openshift-qlvwv-worker-0` which we've created earlier), you can approve it:
+
+```sh
+$ oc adm certificate approve csr-cpnwz
````
For doc consistency, the previous block does an oc describe on csr-lxvbt, but this approve is for csr-cpnwz.  Is that correct?

tomassedovic replied:
Ooops! Fixed in the next commit. Thanks.

tomassedovic commented:
I've realised there's one more thing to do: all the Python scripts should work on Python 3. I think I can make them work on Py2 at the same time without too much trouble but yeah.

So the revised tasks:
- Update Python scripts to work with 3
- Add links
- Resolve the TODO in the manifests section
- Validate the guide still works

tomassedovic commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,806 @@
+```sh
+$ python -c 'import yaml;
+path = "install-config.yaml";
+data = yaml.load(open(path));
````
I've fixed this in a later commit. Should work without warnings now.

tomassedovic commented:
This now has everything I planned to do and it addresses all the requests made by the reviewers so far.

stencell commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,861 @@
+For example, if you want to add the following nameservers: `198.51.100.86` and `198.51.100.87`, you can run this command:
+
+```sh
+$ openstack subnet --dns-nameserver 198.51.100.86 --dns-nameserver 198.51.100.87 "$INFRA_ID-subnet"
````
This command should be `openstack subnet set`, I think.

tomassedovic replied:
Indeed! Thanks for noticing.

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+
+It contrasts with the fully-automated Installer-Provisioned Infrastructure (IPI) which creates everything in one go.
+
+With UPI creating the cloud (OpenStack) resources (e.g. Nova servers, Neutron ports, security groups) is the responsibility of the person deploying OpenShift.
```
I tend to read this line as
```
With UPI creating the cloud, resources [are] the responsibility...
```
instead of
```
With UPI, creating the cloud resources...
```

Consider adding a comma after `UPI`, for readability,.

tomassedovic replied:
Commas have always been the Achilles' heel. Even in my native language \o/.

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+
+The requirements for UPI are broadly similar to the [ones for OpenStack IPI][ipi-reqs]:
+
+[ipi-reqs]: ./openstack#openstack-requirements
```
This expands to
```
https://github.com/openshift/installer/blob/master/docs/user/openstack/openstack#openstack-requirements
```
...which is broken.

Should instead be:
```
https://github.com/openshift/installer/tree/master/docs/user/openstack#openstack-requirements
```

The sanest way to getting there with a relative link could be:
```
[ipi-reqs]:  ./README.md#openstack-requirements
```

pierreprinetti replied:
or poerhaps `../#openstack-requirements`? I have never tried this one

tomassedovic replied:
Thanks!

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+- OpenShift Pull Secret
+- DNS infrastructure you can configure (you'll need to add two records there)
+
+[rhcos]: https://www.openshift.com/learn/coreos/
```
Did you avoid providing a deeper link on purpose?

tomassedovic replied:
I couldn't really find a better one. If you've got suggestions, I'm all ears!

pierreprinetti replied:
> If you've got suggestions

I really haven't! And that's precisely the point: the page reads "RHEL CoreOS is available as part of Red Hat OpenShift", which is a bit of a chicken and egg problem... :thinking:

pierreprinetti replied:
If I even was to find a direct link, I would even not know what are our policies regarding deeplinking Red Hat products.

pierreprinetti replied:
_I have nothing useful to add: please consider as resolved_

tomassedovic replied:
Ok, will do. We can always improve this later.

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+[ipi-reqs]: ./openstack#openstack-requirements
+
+- OpenStack account with `clouds.yaml`
+  - This document will use a cloud called `openstack`
```
Do you think it makes sense to give some visual representation of the two layers of this paragraph ("general case" vs. "this document") by using _italic_ when referring to "this document"?

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+  - This document will use `m1.xlarge` for masters and `m1.large` for workers
+- An external subnet you want to use for floating IP addresses
+  - This document will use `external`
+- The `openshift-install` binary in your `$PATH`
```
or `%PATH%`?

tomassedovic replied:
That's Windows syntax, isn't it? The installer binaries are only available for linux and macos for now. I'm sure there are custom shells that use a different syntax, but `$ENV_VAR_NAME` is the posix-like thing everyone reading this should understand.

pierreprinetti replied:
> `$ENV_VAR_NAME` is the posix-like thing everyone reading this should understand.

I agree 100%. I just thought that a wink to Powershell (or whatever it's called these days) would hint our Windows-based friends that this doc works for them as well. But apparently it doesn't! I wrongly assumed we were providing the Windows binary.

pierreprinetti replied:
_please consider as resolved_

tomassedovic replied:
Gotcha. I wasn't sure myself -- had to check the release page.

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+  - This document will use `external`
+- The `openshift-install` binary in your `$PATH`
+- The [Red Hat CoreOS][rhcos] image in Glance
+  - This document will use the `rhcos` image name
```
```
This document will use `rhcos` as the image name
```
as a newbie, this sounds more explicit to me (we are not asking them to try and find an image named `rhcos` on the Internets)

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+- A base domain
+  - This document will use `example.com`
+- OpenShift Pull Secret
+- DNS infrastructure you can configure (you'll need to add two records there)
```
Do you think we can already be more explicit here?

Maybe something like
```
A DNS zone (or access to `/etc/hosts`) resolvable by the installer machine, for assigning a name to the API external load balancer.
```

or a link to `#create-public-dns-records`, to let the reader figure out what we mean?

pierreprinetti commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,861 @@
+$ openstack flavor show m1.xlarge
+$ openstack flavor show m1.large
+$ which openshift-install
+```
````
this is super useful! Great addition :+1:

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+
+## OpenShift Configuration Directory
+
+All the configuration files, logs and installation state is kept in a single directory:
```
`s/ is / are `

pierreprinetti commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,861 @@
+? Cluster Name <openshift>
+```
+
+Most of these are self-explanatoray. `Cloud` is the cloud name in your `clouds.yaml` i.e. what's set as your `OS_CLOUD` environment variable.
````
I must confess that it is still not entirely clear to me why `Base Domain` is required here, and how it is used by the installer.

Do you think we can add a very short explanation without cluttering the text?

Something like:
```
The fully qualified domain name with which the API interface will expect to the called, and the default name with which OpenShift will expose newly created applications.
```

WDYT?

pierreprinetti commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,861 @@
+open(path, "w").write(yaml.dump(data, default_flow_style=False))'
+```
+
+**NOTE**: All the script in this guide work with Python 3 as well as Python 2.
````
script+s

tomassedovic commented:
I've verified that following this latest version results in a successful UPI deployment.

I've also addressed all the comments I could.

@pierreprinetti please reply to the two outstanding ones.

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,861 @@
+Our range will be `192.0.2.0/24` so we need to add that value to
+`install-config.yaml`. Look under `networking` -> `machineCIDR`.
+
+This command will do it for you:
```
General question, please just ignore if already discussed or out of scope.

I read that [`install-config.yaml` is part of the stable Installer API](https://github.com/openshift/installer/blob/master/docs/user/overview.md#multiple-invocations). Wouldn't it be more straightforward to provide here an `install-config.yaml` template where to fill IPs and CIDRs and flavours by hand, rather than asking the users to feed `openshift-install` interactively and _then_ change the output YAML?

Personally, as soon as I figured out that I could fine-tune my cluster by editing `install-config.yaml`, I have never come back again to `create cluster` without providing a YAML.

tomassedovic replied:
Possibly, but it's what AWS UPI does as well. I didn't want to strike on our own on things like that: https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#create-configuration

pierreprinetti commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,871 @@
+
+To provide access to the OpenShift cluster, we will need to create two ports, attach our API and Ingress Floating IPs and publish the appropriate DNS records.
+
+**IMPORTANT:** these ports will need to have specific fixed (i.e. private) IP addresses. They have to be the fifth and seventh addresses from the subnet range. For example:
```
Why does `fixed == private` here? (I might be wrong but that's my interpretation of "id est")

tomassedovic replied:
In OpenStack terminology, the private IP address (only addressable within the subnet the server/port is) is a `Fixed IP`.

In most classic OpenStack deployments (where you need routers and floating IPs to provide connectivity beyond the subnet) fixed == private and floating == public.

But: there are deployments (the key term is generally "provider networks") where you don't need floating IPs and any server you create will have its fixed IP also accessible from the outside.

tomassedovic commented on docs/user/openstack/install_upi.md:
```diff
@@ -0,0 +1,871 @@
+
+### Update Bootstrap Ignition
+
+There is some bootstrap configuration we weed to add explicitly to the ignition file. The IPI installer does this automatically via Terraform, but the changes do not show up in the files generated by `create ignition-configs`. This section should go away in the future.
```
It is to the end-users: someone reading this might wonder: why am I manually making these ignition changes here? They're not necessary for the AWS UPI and they're done automatically with the OpenStack IPI. So what gives?

Maybe I could word this better?

pierreprinetti replied:
OK I am fine with addressing this nitpick at a later stage (possibly, when we actually get rid of what we promised we get rid of :grin: )

mandre commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,871 @@
+
+```sh
+$ openstack floating ip delete 203.0.113.23 203.0.113.19
+$ openstack server list --column Name --format value | grep "$INFRA_ID" | xargs openstack server delete
````
Could you add `--no-run-if-empty` to xargs so that it doesn't fail when there is no matching server?

mandre commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,871 @@
+```sh
+$ openstack floating ip delete 203.0.113.23 203.0.113.19
+$ openstack server list --column Name --format value | grep "$INFRA_ID" | xargs openstack server delete
+$ openstack port list --column Name --format value | grep "$INFRA_ID" | xargs openstack port delete
````
ditto

tomassedovic commented on docs/user/openstack/install_upi.md:
````diff
@@ -0,0 +1,871 @@
+
+```sh
+$ openstack network create "$INFRA_ID-network"
+$ openstack subnet create --subnet-range <192.0.2.0/24> --allocation-pool start=192.0.2.10,end=192.0.2.254 --network "$INFRA_ID-network" "$INFRA_ID-subnet"
````
I've put them in places where people will definitely want to edit it. This subnet is the documentation one and it shouldn't actually be used by anyone (according to the RFC, firewalls etc. are encouraged to drop packets with it).

tomassedovic commented:
I've squashed the commits. I believe this is ready to go. Obviously if anyone has any other significant requests, I'll be happy to acomodate them.

That said, commits like this can get easily mired in small change requests -- and we can always just open up new PRs. So I'd prefer to merge this and then improve it later if that's fine with everyone.

pierreprinetti commented:
/lgtm
Great job!

tomassedovic commented:
Thanks!
